keras/preprocessing/sequence.py

```diff
@@ -216,3 +218,107 @@ def _remove_long_seq(maxlen, seq, label):
             new_seq.append(x)
             new_label.append(y)
     return new_seq, new_label
+
+
+class TimeseriesGenerator(Sequence):
+    """Utility class for generating batches of temporal data.
+
+    This class takes in a sequence of data-points gathered at
+    equal intervals, along with time series parameters such as
+    stride, length of history etc., to produce batches for
+    training/validation.
+
+    # Arguments
+        data: indexable generator (such as list or Numpy array)
+            containing consecutive data points.
+        targets: targets corresponding to instances in `data`.
+            Should have same length as `data`.
+        length: length of history to consider for prediction.
+            `data[i-length : i+1]` is used to prepare the input.
+        sampling_rate: period between successive individual data
+            points. For rate `r`, points
+            `data[i]`, `data[i-r]`, ... `data[i - length]`
+            are used as input.
+        stride: period between successive sub-sequences used for
+            training. For stride `s`, consecutive training samples would
+            be centered around `data[i]`, `data[i+s]`, `data[i+2*s]`, ...
+        start_index, end_index: delimiters for sub-sets of `data` & `targets`
+            to be used for batch generation. Useful to reserve part of
+            data for validation.
+        shuffle: whether to shuffle training samples, or draw them in
+            chronological order.
+        reverse: if `true`, input points are provided in reverse order.
+        batch_size: number of timeseries samples in each batch
+            (except maybe last one)
+
+    # Returns
+        A `Sequence` instance.
+
+    # Examples
+
+    ```data = np.array([[i] for i in range(50)])
+       targets = np.array([i for i in range(50)])
+
+       data_gen = TimeseriesGenerator(data, targets,
+                                      length=10, sampling_rate=2,
+                                      batch_size=2)
+       assert len(data_gen) == 20
+       assert (np.array_equal(data_gen[0][0],
+                              np.array([[[0], [2], [4], [6], [8]],
+                                        [[1], [3], [5], [7], [9]]])))
+       assert (np.array_equal(data_gen[0][1],
+                              np.array([[10], [11]])))
+    ```
+    """
+
+    def __init__(self, data, targets,
+                 length,
+                 sampling_rate=1,
+                 stride=1,
+                 start_index=0,
+                 end_index=None,
+                 shuffle=False,
+                 reverse=False,
+                 batch_size=128):
+        self.data = data
+        self.targets = targets
+        if not hasattr(self.targets[0], '__len__'):
+            self.output_dim = 1
+        else:
+            self.output_dim = len(self.targets[0])
```

Comment:
These few lines seem to restrict `targets` to being 1D. In practice you may want to have more complicated targets. One concrete example is trying to predict a segmentation mask for a video frame (potentially blurry) given a timeseries of previous frames up to that point.

It sounds like you should not store `output_dim` at all, and rather rely on `targets.shape` when constructing the batch targets.